libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/target/SoqlSyncDownTarget.kt

```diff
@@ -0,0 +1,256 @@
+    override fun asJSON(): JSONObject {
+        return with(super.asJSON()) {
+            put(QUERY, query)
+            put(MAX_BATCH_SIZE, maxBatchSize)
+        }
+    }
+
+    @Throws(IOException::class, JSONException::class)
+    override fun startFetch(syncManager: SyncManager, maxTimeStamp: Long): JSONArray? {
+        return startFetch(syncManager, getQuery(maxTimeStamp))
+    }
+
+    @Throws(IOException::class, JSONException::class)
+    protected fun startFetch(syncManager: SyncManager, query: String): JSONArray {
+        val request = RestRequest.getRequestForQuery(syncManager.apiVersion, query, maxBatchSize)
+        val response = syncManager.sendSyncWithMobileSyncUserAgent(request)
+        val responseJson = getResponseJson(response)
+        val records = getRecordsFromResponseJson(responseJson)
+
+        // Records total size.
+        totalSize = responseJson.getInt(Constants.TOTAL_SIZE)
+
+        // Captures next records URL.
+        nextRecordsUrl = JSONObjectHelper.optString(responseJson, Constants.NEXT_RECORDS_URL)
+        return records
+    }
+
+    @Throws(IOException::class)
+    protected fun getResponseJson(response: RestResponse): JSONObject {
+        val responseJson: JSONObject = try {
+            response.asJSONObject()
+        } catch (e: JSONException) {
+            // Rest API errors are returned as JSON array
+            throw MobileSyncException(response.asString())
+        }
+        return responseJson
+    }
+
+    @Throws(JSONException::class)
+    protected open fun getRecordsFromResponseJson(responseJson: JSONObject): JSONArray {
+        return responseJson.getJSONArray(Constants.RECORDS)
+    }
+
+    @Throws(IOException::class, JSONException::class)
+    override fun continueFetch(syncManager: SyncManager): JSONArray? {
+        if (nextRecordsUrl == null) {
+            return null
+        }
+        val request = RestRequest(RestRequest.RestMethod.GET, nextRecordsUrl)
+        val response = syncManager.sendSyncWithMobileSyncUserAgent(request)
+        val responseJson = getResponseJson(response)
+        val records = getRecordsFromResponseJson(responseJson)
+
+        // Captures next records URL.
+        nextRecordsUrl = JSONObjectHelper.optString(responseJson, Constants.NEXT_RECORDS_URL)
+        return records
+    }
+
+    @Throws(IOException::class, JSONException::class)
+    override fun getRemoteIds(syncManager: SyncManager, localIds: Set<String>): Set<String> {
+        return getRemoteIdsWithSoql(syncManager, soqlForRemoteIds)
+    }
+
+    @Throws(IOException::class, JSONException::class)
+    protected fun getRemoteIdsWithSoql(
+        syncManager: SyncManager,
+        soqlForRemoteIds: String
+    ): Set<String> {
+        val remoteIds: MutableSet<String> = HashSet()
+
+        // Makes network request and parses the response.
+        remoteIds.addAll(parseIdsFromResponse(startFetch(syncManager, soqlForRemoteIds)))
+        while (true) {
```

Comment:
@wmathurin - Is this intended to be `while (true)`? The previous code had a check on `records != null`.

Reply:
startFetch always return something not null while continue fetch can return null.
The full loop does:
``` kotlin
        while (true) {
            syncManager.checkAcceptingSyncs()

            // Fetch next records, if any.
            val records = continueFetch(syncManager) ?: break
            remoteIds.addAll(parseIdsFromResponse(records))
        }
```
Another Elvis-break - maybe not the most readable code?

Reply:
Before it did
```java
while (records != null) {
            syncManager.checkAcceptingSyncs();

            // Fetch next records, if any.
            records = continueFetch(syncManager);
            remoteIds.addAll(parseIdsFromResponse(records));
        }
```

Reply:
I'll rewrite for clarity.

Reply:
See https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/2459/commits/985768f21d26146683e7f8efb342680e6b604c6d